Source/Urho3D/Resource/GraphPin.h

```diff
@@ -0,0 +1,154 @@
+// THE SOFTWARE.
+//
+
+#pragma once
+
+#include "../Core/Context.h"
+#include <EASTL/fixed_vector.h>
+
+namespace Urho3D
+{
+class GraphNode;
+template <typename T, size_t nodeCount> struct GraphNodeMapHelper;
+
+enum GraphPinDirection
+{
+    PINDIR_INPUT,
+    PINDIR_OUTPUT,
+    PINDIR_ENTER,
+    PINDIR_EXIT,
+};
+
+/// Abstract graph node pin.
+class URHO3D_API GraphPin
+{
+protected:
+    /// Construct.
+    GraphPin(GraphNode* node, GraphPinDirection direction);
+
+public:
+    /// Get pin type.
+    /// @property 
+    VariantType GetType() const { return type_; }
+
+    /// Set pin type.
+    /// @property
+    void SetType(VariantType type);
+
+    /// Get name of the pin.
+    /// @property
+    const ea::string& GetName() { return name_; }
+
+    /// Get pin direction.
+    GraphPinDirection GetDirection() const { return direction_; }
+
+    /// Get pin node.
+    GraphNode* GetNode() const { return node_; }
+
+protected:
+    /// Serialize from/to archive. Return true if successful.
+    virtual bool Serialize(Archive& archive);
+
+    /// Set name of the pin. Executed by GraphNode.
+    /// @property
+    void SetName(const ea::string_view name);
+
+    /// Pin name.
+    ea::string name_;
+
+    /// Pin type.
+    VariantType type_;
+
+
+    GraphNode* node_;
+    GraphPinDirection direction_;
+
+    friend class GraphNode;
+    template <typename T, size_t nodeCount> friend class GraphNodeMapHelper;
+};
+
+/// Graph node pin that connects to other pins.
+class URHO3D_API GraphOutPin : public GraphPin
+{
+protected:
+    /// Construct.
+    GraphOutPin(GraphNode* node, GraphPinDirection direction);
+
+    friend class GraphNode;
+};
+
+/// Graph node pin with connection.
+class URHO3D_API GraphInPin : public GraphPin
+{
+protected:
+    /// Construct.
+    GraphInPin(GraphNode* node, GraphPinDirection direction);
+
+    /// Serialize from/to archive. Return true if successful.
+    bool Serialize(Archive& archive) override;
+
+public:
+    /// Connect to other pin.
+    bool ConnectTo(GraphOutPin& pin);
+
+    /// Get pin connected to the current pin.
+    GraphOutPin* GetConnectedPin() const;
+
+    /// Get true if pin is connected.
+    bool GetConnected() const { return targetNode_; }
+
+private:
+    /// Target node.
+    unsigned targetNode_;
+
+    /// Target pin name.
+    ea::string targetPin_;
+
+    friend class GraphNode;
+};
+
+/// Graph data flow node pin with connection and default value.
+class URHO3D_API GraphDataInPin : public GraphInPin
+{
+protected:
+    /// Construct.
+    GraphDataInPin(GraphNode* node, GraphPinDirection direction);
+
+    /// Serialize from/to archive. Return true if successful.
+    bool Serialize(Archive& archive) override;
+
+public:
+    /// Get default value.
+    /// @property
+    const Variant& GetDefaultValue();
+
+    /// Set default value.
+    /// @property 
+    void SetDefaultValue(const Variant& variant);
+
+private:
+    /// Target node.
+    Variant defaultValue_;
```

Comment:
Any reason to have separate GraphDataInPin and GraphInPin?

Reply:
Yes. Data In pin can have a value attached. I guess In pin should be called differently. Or maybe I should not mix then at all.